KAFKA-15778 & KAFKA-15779: Implement metrics manager (KIP-714)

apoorvmittal10 opened the pull request:
The PR provide implementation for client metrics manager along with other classes. Manager is responsible to support 3 operations:

1. UpdateSubscription - From `kafka-configs.sh` and reload from metadata cache.
2. Process Get Telemetry Request - From KafkaApis.scala
3. Process Push Telemetry Request - From KafkaApis.scala

Manager maintains an in-memory cache to keep track of client instances against their instance id.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

apoorvmittal10 committed:
KAFKA-15778 & KAFKA-15779: Implement metrics manager with client instance state (KIP-714)

apoorvmittal10 commented on core/src/main/java/kafka/metrics/ClientMetricsConfigs.java:
```diff
@@ -0,0 +1,194 @@
+ * For more information please look at kip-714:
+ * https://cwiki.apache.org/confluence/display/KAFKA/KIP-714%3A+Client+metrics+and+observability#KIP714:Clientmetricsandobservability-Clientmetricsconfiguration
+ */
+public class ClientMetricsConfigs extends AbstractConfig {
```
This class is major overlap from PR where it was introduced: https://github.com/apache/kafka/pull/14621

apoorvmittal10 replied:
Resolved conflicts after merging the upstream.

apoorvmittal10 commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -0,0 +1,416 @@
+        return INSTANCE;
+    }
+    // Max cache size (16k active client connections per broker)
+    private static final int CM_CACHE_MAX_SIZE = 16384;
```
The cache should hold the information of all connected client hence it should support the number of clients that can connect to single broker. Kept it high enough for now but would require suggestions for best approach. @AndrewJSchofield mentioned that it might be the `max.connections` broker config but the configs uppoer bound is  Int.MAX_VALUE hence we need to think about better upper bound here.

junrao replied:
Is this enough? A broker could handle 100s of Ks of connections.

apoorvmittal10 replied:
Then should it be in low 100K i.e. 2^17 = 131072? I am not sure what the right config should be here, I know the cloud providers like MSK typically provides 3000 active connections per broker where this limit is higher in case of Confluent but what typically are the active number of client connecting to broker should be.

junrao replied:
The challenge is that the connections can vary depending on the type of broker host. Larger instance typically can accommodate more connections. Do we need to make this an LRU cache? If a client is terminated or idle, we should remove them from the cache. Otherwise, we probably should just rely on the existing `max.connections `to control the client connections?

apoorvmittal10 replied:
> Do we need to make this an LRU cache? If a client is terminated or idle, we should remove them from the cache

I have started with LRU cache and planning to improve this with cache which timebounds the connection. The KIP says: `client instance specific state is maintained in broker memory up to MAX(60*1000, PushIntervalMs * 3) milliseconds`. I ll add improvement on the cache to respect that: https://issues.apache.org/jira/browse/KAFKA-15813

apoorvmittal10 commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -0,0 +1,416 @@
+    }
+    // Max cache size (16k active client connections per broker)
+    private static final int CM_CACHE_MAX_SIZE = 16384;
+    private final Cache<Uuid, ClientMetricsInstance> clientInstanceCache;
```
The KIP-714 says: `This client instance specific state is maintained in broker memory up to MAX(60*1000, PushIntervalMs * 3) milliseconds and is used to enforce the push interval rate-limiting. There is no persistence of client instance metrics state across broker restarts or between brokers.`

I have started with LRUCache with oldest entry eviction but will implement something similar to `Selector.IdleExpiryManager` which cleansup old connections.

apoorvmittal10 committed:
Adding kafka config with telemerty config

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into kip-714-ak-manager

apoorvmittal10 commented on core/src/main/java/kafka/metrics/ClientMetricsConfigs.java:
```diff
@@ -143,46 +152,43 @@ private static void validateProperties(Properties properties) {
-                    throw new InvalidConfigurationException("Illegal client matching pattern: " + pattern);
-                }
-            });
+    public static Map<String, Pattern> parseMatchingPatterns(List<String> patterns) {
```
Moved from <String, String> to <String, Pattern> to avoid reparsing of patterns.

apoorvmittal10 committed:
Updated request time tracking

apoorvmittal10 committed:
Update subscribed metrics config to include all

apoorvmittal10 commented:
@junrao @hachikuji @AndrewJSchofield @mjsax Please if I can get feedback on the PR.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/PushTelemetryRequest.java:
```diff
@@ -71,6 +73,31 @@ public PushTelemetryRequestData data() {
+        return new PushTelemetryResponse(responseData);
+    }
+
+    public String getMetricsContentType() {
```
We don't use getters. So this can just be `metricsContentType`. Ditto for `getMetricsData`.

apoorvmittal10 replied:
Missed these, thanks. Done.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -16,12 +16,48 @@
  */
 package kafka.server;
 
+import java.util.Collections;
```
Should this class be in the same package as other client metric related classes like `DefaultClientTelemetryPayload`?

apoorvmittal10 replied:
The reason I kept this at `kafka.server` package as I see all managers (in scala) processing API calls from KafkaApis.scala resides `kafka.server` package.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -34,13 +70,348 @@ public class ClientMetricsManager implements Closeable {
     @Override
     public void close() throws IOException {
-        // TODO: Implement the close logic to close the client metrics manager.
+        // Do nothing for now.
```
Should we clear up the internal data?

apoorvmittal10 replied:
I have cleaned up the subscription map, there is no method currently exposed to clear cache but I ll add that as improvement while improving the cache code.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -34,13 +70,348 @@ public class ClientMetricsManager implements Closeable {
+
+    // The last subscription updated time is used to determine if the next telemetry request needs
+    // to re-evaluate the subscription id as per changes subscriptions.
+    private long lastSubscriptionUpdateEpoch;
```
Should this be volatile since it's written and read by different threads?

apoorvmittal10 replied:
Make sense, thanks. Done.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -34,13 +70,348 @@ public class ClientMetricsManager implements Closeable {
+            ClientMetricsInstanceMetadata instanceMetadata = new ClientMetricsInstanceMetadata(
+                clientInstanceId, requestContext);
+            clientInstance = createClientInstanceAndUpdateCache(clientInstanceId, instanceMetadata, timestamp);
+        } else if (clientInstance.subscriptionUpdateEpoch() < lastSubscriptionUpdateEpoch) {
```
Hmm, a subscription could be updated again in the same millisec. Is this check sufficient?

apoorvmittal10 replied:
As I do not think that subscription creation/updation would be too frequent i.e. number of subscriptions should be fairly low in broker hence the situation where client instance is created at the same time of subscription should be rare. But when a broker is re-started and there are clients waiting for the connection, then subscription ll be loaded (from dynamic config handler) at the same time when instance is created hence I skipped `=` check to avoid re-evaluation of subscriptions.

 Also in the next telemetry API call re-evaluation of update timestamp will happen which shall either update the subscription or throw appropriate error to client.

junrao replied:
Hmm, is skipping the `=` check enough? It seems that it's possible that a client picks up the first subscription, but misses the second subscription. If that happens, it's possible that a client won't pick up a newly matched subscription for a long time, right? Could we just use a counter and bump it up each time when there is a subscription update?

apoorvmittal10 replied:
Yes you are right, that could be a problem. But using atomic integer can bump the version appropriately but we need a definite way to track the subscription evaluation version in client instance. Hence either we need to have read and write locks to store correct subscription version in client instance or, the approach I took in the PR, store the version with each subscription info and while evaluating subscriptions for client instance, store the maximum version encountered.

I have addressed the changes required, done.

apoorvmittal10 replied:
Moved to single global variable, resolving the comment.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -34,13 +70,348 @@ public class ClientMetricsManager implements Closeable {
+
+        for (SubscriptionInfo info : subscriptions()) {
+            if (instanceMetadata.isMatch(info.matchPattern())) {
+                allMetricsSubscribed = allMetricsSubscribed || info.metrics().contains(
```
We don't need to include `allMetricsSubscribed` since it's false at this point?

apoorvmittal10 replied:
I think I might be missing something here to understand. Though the `allMetricsSubscribed` will be false in first iteration but as the subscriptions has to be iterated completely anyways to fetch minimum `interval.ms` hence I cannot break from the iteration when `allMetricsSubscribed` becomes `true`. The other way of writing the code could be below, but it's effectively same.

```
if (!allMetricsSubscribed) {
  allMetricsSubscribed = info.metrics().contains(ClientMetricsConfigs.ALL_SUBSCRIBED_METRICS_CONFIG);
}
```

junrao replied:
Thanks for the explanation, @apoorvmittal10. You are right. So, this is not an issue.

junrao commented on core/src/main/java/kafka/metrics/ClientMetricsInstance.java:
```diff
@@ -0,0 +1,121 @@
+    private boolean terminating;
+    private long lastGetRequestEpoch;
+    private long lastPushRequestEpoch;
+    private Errors lastKnownError;
```
Does this need to be volatile since it's written and read by different threads. Ditto for other fields like `terminating`

apoorvmittal10 replied:
I have marked 2 of them volatile and other 2 are protected now with `synchronized`. I have added additional concurrency tests in to validate the behaviour that a single request can process when multiple are received for same client instance, others get rejected by throttling error.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -34,13 +70,348 @@ public class ClientMetricsManager implements Closeable {
+        }
+    }
+
+    private List<Byte> getSupportedCompressionTypes() {
```
This can just be `supportedCompressionTypes`.

apoorvmittal10 replied:
Done.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -34,13 +70,348 @@ public class ClientMetricsManager implements Closeable {
+
+    private List<Byte> getSupportedCompressionTypes() {
+        List<Byte> compressionTypes = new ArrayList<>();
+        for (CompressionType compressionType : CompressionType.values()) {
```
Are we returning the compression type in the preferred order?

apoorvmittal10 replied:
Thanks for pointing that out, I missed that. I have moved it to static map which follows the order defined in the KIP.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -34,13 +70,348 @@ public class ClientMetricsManager implements Closeable {
+
+    private ClientMetricsInstance getClientInstance(Uuid clientInstanceId, RequestContext requestContext,
+        long timestamp) {
+        // Check if null can be called on the cache. if can then we can avoid the method call.
```
Have you checked that yet?

apoorvmittal10 replied:
Sorry, removed the comment. Yes I did check it :)

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -34,13 +70,348 @@ public class ClientMetricsManager implements Closeable {
+        if (clientInstanceId == null || clientInstanceId == Uuid.ZERO_UUID) {
+            String msg = String.format("Invalid request from the client [%s], invalid client instance id",
+                clientInstanceId);
+            return request.getErrorResponse(throttleMs, new InvalidRequestException(msg));
```
Do we need to set `throttleMs` for invalid request error?

apoorvmittal10 replied:
I have asked the same question in other comment but shouldn't we set that as API call goes through `sendMaybeThrottle` in KafkaApis which throttles the request?

apoorvmittal10 replied:
I have removed the throttle time THROTTLING_QUOTA_EXCEEDED as per the explanation in other comment.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/PushTelemetryRequest.java:
```diff
@@ -71,6 +73,31 @@ public PushTelemetryRequestData data() {
+    public PushTelemetryResponse createResponse(int throttleTimeMs, Errors errors) {
+        PushTelemetryResponseData responseData = new PushTelemetryResponseData();
+        responseData.setErrorCode(errors.code());
+        responseData.setThrottleTimeMs(throttleTimeMs);
```
If error code is not `ThrottlingQuotaExceededException`, we should ignore `throttleTimeMs`. 

Could we redirect `getErrorResponse` to here and rename it to `errorResponse`?

apoorvmittal10 replied:
> Could we redirect getErrorResponse to here and rename it to errorResponse?

Done

> If error code is not ThrottlingQuotaExceededException, we should ignore throttleTimeMs.

It might be naive but sorry I didn't understand as why throttleTimeMs should not be passed in response for other exceptions. Isn't all requests goes through common throttling code where requests might be throttled for sometime based on the throughput?

junrao replied:
The following is my understanding. There are two types of throttling.

1. Generic request throttling based on % of CPU a client uses on the broker. This applies to any request, including `PushTelemetry`. In this case, the error code is none and the `throttleTimeMs` field is set. The client will mute the channel for `throttleTimeMs` before sending future requests.
2. `PushTelemetry` specific throttling because PushTelemetry is sent too frequently. In this case, we should set THROTTLING_QUOTA_EXCEEDED as the error code and avoid setting the `throttleTimeMs` field since we don't want to the client to mute the channel for all requests.

apoorvmittal10 replied:
Thanks for the explanation @junrao, this is helpful. I have removed throttleTimeMs for the case when THROTTLING_QUOTA_EXCEEDED exception is thrown in telemetry APIs. Done.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -34,13 +70,348 @@ public class ClientMetricsManager implements Closeable {
+            return request.getErrorResponse(throttleMs, exception);
+        } finally {
+            // Update the client instance with the latest push request parameters.
+            clientInstance.terminating(request.data().terminating());
```
The KIP didn't talk about this. But if the client is hard killed, should we expire the client instance on the broker after some time?

apoorvmittal10 replied:
The KIP says that it should be `MAX(60*1000, PushIntervalMs * 3) milliseconds`, I ll add that with cache improvement.

apoorvmittal10 committed:
Adding compression type error code for retry

junrao commented on core/src/main/java/kafka/metrics/ClientMetricsReceiverPlugin.java:
```diff
@@ -0,0 +1,52 @@
+ */
+public class ClientMetricsReceiverPlugin {
+
+    private static final List<ClientTelemetryReceiver> RECEIVERS = new ArrayList<>();
```
Could we make the field and the method instance level instead of static?

apoorvmittal10 replied:
Done.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -34,13 +70,348 @@ public class ClientMetricsManager implements Closeable {
+        // Push the metrics to the external client receiver plugin.
+        byte[] metrics = request.data().metrics();
+        if (metrics != null && metrics.length > 0) {
+            ClientMetricsReceiverPlugin.exportMetrics(requestContext, request);
```
Should we catch any exception and convert that to an `INVALID_RECORD` in the response?

apoorvmittal10 replied:
Done.

junrao commented on core/src/main/java/kafka/metrics/ClientMetricsInstanceMetadata.java:
```diff
@@ -0,0 +1,72 @@
+        // broker's point of view. But the broker does not have this information rather the port could be
+        // the broker's port where the client connection is established. We might want to consider removing
+        // the client source port from the KIP or use broker port if that can be helpful.
+        // TODO: fix port
```
We populate `RequestContext` in `SocketServer`. We could obtain the client port through `transportLayer.socketChannel().socket().getPort` in `KafkaChannel`.

apoorvmittal10 replied:
Thanks a lot @junrao this is helpful. I will make the changes in subsequent PR to address this. I have created following jira for same: https://issues.apache.org/jira/browse/KAFKA-15811

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/PushTelemetryRequest.java:
```diff
@@ -71,6 +73,31 @@ public PushTelemetryRequestData data() {
+    }
+
+    private static ByteBuffer decompressMetricsData(CompressionType compressionType, byte[] metrics) {
+        // TODO: Add support for decompression of metrics data
```
So, this will be added in a future PR?

apoorvmittal10 replied:
Yes, I have created jira in parent KIP-714 task to address this: https://issues.apache.org/jira/browse/KAFKA-15807. I am planning to get end-to-end metrics flow without compression first.

junrao commented on core/src/main/java/kafka/metrics/ClientMetricsConfigs.java:
```diff
@@ -80,6 +82,11 @@ public class ClientMetricsConfigs {
     public static final String CLIENT_SOURCE_PORT = "client_source_port";
 
+    // Empty string in client-metrics resource configs indicates that all the metrics are subscribed.
+    public static final String ALL_SUBSCRIBED_METRICS_CONFIG = "\"\"";
```
This is what the user provides. Why is `ALL_SUBSCRIBED_METRICS` not the same as this?

apoorvmittal10 replied:
The KIP mentions that the response in subscriptions should be just empty string i.e. `""`. But to create `client-metrics` resource the ConfigDef parses `""` as no data hence to specify empty string through `kafka-configs.sh` we need to pass them as empty string enclosed in string i.e. `"\"\""`

junrao replied:
Hmm, this can be confusing. Would it be better to change the KIP to use sth like `*` to represent all metrics? cc @AndrewJSchofield

AndrewJSchofield replied:
It would be possible to change the KIP to support a list containing just `*` to mean all metrics. The KIP today says a list containing just `""` means all metrics. That's a pretty simple change.

apoorvmittal10 replied:
I have moved to '*' as the identifier, done.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the PR. Made a pass of non-testing files. Left a few comments.

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 commented:
> @apoorvmittal10 : Thanks for the PR. Made a pass of non-testing files. Left a few comments.

Thanks a lot for the review @junrao . I have addressed the comments and have a question related to throttleTimeMs for errors in the comments. Please if you can re-review.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/PushTelemetryRequest.java:
```diff
@@ -60,17 +62,39 @@ public PushTelemetryRequest(PushTelemetryRequestData data, short version) {
-                .setErrorCode(Errors.forException(e).code())
-                .setThrottleTimeMs(throttleTimeMs);
-        return new PushTelemetryResponse(responseData);
+        return errorResponse(throttleTimeMs, Errors.forException(e));
```
This is an existing issue, but `getErrorResponse` can just be `errorResponse`.

apoorvmittal10 replied:
I might have missed something here to understand, the method to override has to be `getErrorResponse` and I just created another method `errorResponse` which helps building the response. Is there anything I need to address here?

junrao replied:
Got it. Since getErrorResponse is an override of a base method, we can leave it as it is.

junrao commented on core/src/main/java/kafka/metrics/ClientMetricsReceiverPlugin.java:
```diff
@@ -0,0 +1,62 @@
+ */
+public class ClientMetricsReceiverPlugin {
+
+    private static final ClientMetricsReceiverPlugin INSTANCE = new ClientMetricsReceiverPlugin();
```
static fields like that could make it more difficult to test multiple broker instances in the same JVM. Could we create an instance when creating a broker and pass that instance along? Ditto for `ClientMetricsManager.INSTANCE`.

apoorvmittal10 replied:
Make sense, for `ClientMetricsManager` the class is needed to be in BrokerServer.scala and ControllerServer.scala as the client-metrics resource reload has to be managed at a singleton instance. Shall I add the changes in SharedServer.scala to share the same class?

For ClientMetricsReceiverPlugin, here is the PR that uses same https://github.com/apache/kafka/pull/14767/files#diff-019dd06900a2f4714d626ffb071466765413970462f36a86e8a37c0ef9c0afb9, as the class is required while defining DynamicBrokerConfig hence I thought to make it singleton as well.

junrao replied:
Hmm, why is ClientMetricsManager needed in the controller? I thought only brokers can receive get/push telemetric requests.

apoorvmittal10 replied:
Yeah, I re-checked the code today and we shouldn't require any change in ControllerServer.scala as the APIs are not supported on controller and we won't require even KIP-1000 to be supported on controllers. I am making the change to initialize ClientMetricsManager only from KafkaServer.scala

What do you think about `ClientMetricsReceiverPlugin` based on the usage in the tagged PR above?

apoorvmittal10 replied:
I have moved the `ClientMetricsManager` static instance and removed the code from ControllerServer.scala, now initialization only happens in `BrokerServer.scala`. Please let me know your thought for `ClientMetricsReceiverPlugin`.

junrao replied:
For `ClientMetricsReceiverPlugin`, it seems that you could do the same thing by creating an instance in `BrokerServer` and pass it along to `DynamicBrokerConfig`.

apoorvmittal10 replied:
Done.

junrao commented on core/src/test/java/kafka/metrics/ClientMetricsInstanceMetadataTest.java:
```diff
@@ -0,0 +1,136 @@
+            ClientMetricsTestUtils.requestContext());
+        // We consider empty/missing client matching patterns as valid
+        assertTrue(instanceMetadata.isMatch(Collections.emptyMap()));
+        assertTrue(instanceMetadata.isMatch(null));
```
How could a user create a null pattern?

AndrewJSchofield replied:
Good question. `kafka-client-metrics.sh .... --alter --name CM1 --match ""` deletes the map.

apoorvmittal10 replied:
Make sense, user cannot. I have removed this test condition and null check in the code as well.

junrao commented on core/src/test/java/kafka/metrics/ClientMetricsInstanceTest.java:
```diff
@@ -0,0 +1,104 @@
+    public void setUp() throws UnknownHostException {
+        uuid = Uuid.randomUuid();
+        instanceMetadata = new ClientMetricsInstanceMetadata(uuid,
+            ClientMetricsTestUtils.requestContext());
```
merge with previous line?

apoorvmittal10 replied:
I have used `uuid` var later as well to check if `match` is successful hence kept it separate. But if it's for merging the line 37 and 38 then I did that.

junrao replied:
Yes, I just meant merging line 37 and 38.

junrao commented on core/src/test/java/kafka/metrics/ClientMetricsInstanceTest.java:
```diff
@@ -0,0 +1,104 @@
+            null, 2);
+        assertTrue(clientInstance.maybeUpdateGetRequestEpoch(System.currentTimeMillis()));
+        // sleep for 3 ms to ensure that the next request is accepted.
+        Thread.sleep(3);
```
Could we use MockTime to avoid actual sleeping during tests?

apoorvmittal10 replied:
I have changed the code so we don't require sleep in tests, done. Also started using MockTime in ClientMetricsManager.

junrao commented on core/src/test/java/kafka/metrics/ClientMetricsInstanceTest.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    @Test
+    public void testMaybeUpdatePushRequestAfterElapsedTimeValid() throws InterruptedException {
```
How is this test different from `testMaybeUpdateGetRequestAfterElapsedTimeValid`?

apoorvmittal10 replied:
The methods independently check for updating getTelemetryTime and pushTelemetryTime respectively. Is the concern more about that we can merge them in single test? As I have removed the sleep now from methods then do you think it's fine to keep them separate?

junrao commented on core/src/test/java/kafka/metrics/ClientMetricsReceiverPluginTest.java:
```diff
@@ -0,0 +1,56 @@
+
+        assertEquals(1, telemetryReceiver.exportMetricsInvokedCount);
+        assertEquals(1, telemetryReceiver.metricsData.size());
+        assertEquals("test-metrics", new String(
```
Could we just directly compare the bytes?

apoorvmittal10 replied:
Done.

junrao commented on core/src/test/java/kafka/server/ClientMetricsManagerTest.java:
```diff
@@ -0,0 +1,921 @@
+        clientMetricsManager.updateSubscription("sub-1", new Properties());
+        // No subscription should be added as the properties are empty.
+        assertEquals(0, clientMetricsManager.subscriptions().size());
+        assertEquals(previousEpoch, clientMetricsManager.lastSubscriptionUpdateEpoch());
```
`clientMetricsManager.lastSubscriptionUpdateEpoch()` could  change during the test, right? Could we use MockTime to make the test more reliable?

apoorvmittal10 replied:
`clientMetricsManager` is reset to new ClientMetricsManager before each test so parallel execution of tests should not affect individual tests. In this case, we are testing that empty properties request should not make any changes 
 in the subscription and subscription update time as empty properties request should be ignored if no existing subscription with same name exists. Please let me know if I missed anything here.

apoorvmittal10 replied:
I have started using MockTime and subscriptionUpdateVersion AtomicCounter, done.

junrao commented on core/src/test/java/kafka/server/ClientMetricsManagerTest.java:
```diff
@@ -0,0 +1,921 @@
+        properties.put("random", "random");
+        clientMetricsManager.updateSubscription("sub-1", properties);
+
+        // Undefined keys are ignored.
```
Hmm, this is kind of inconsistent. If we remove all properties in a config, we end up removing the subscription. However, if we create a subscription with a non-existing property, we end up with a subscription with the default properties.

apoorvmittal10 replied:
Though undefined keys shouldn't arrive in this method as pre-validation happens while processing data from ConfigCommand. I have added a validation check in this method as well and changed the test to see correct exception is thrown.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -30,17 +68,366 @@ public class ClientMetricsManager implements Closeable {
+     * the PushIntervalMs, XORed with the ClientInstanceId.
+     */
+    private int computeSubscriptionId(Set<String> metrics, int pushIntervalMs, Uuid clientInstanceId) {
+        CRC32 crc = new CRC32();
```
We have moved to CRC32C since it's more efficient (https://issues.apache.org/jira/browse/KAFKA-1449). Should we use CRC32C here too?

apoorvmittal10 replied:
I have moved to CRC32C, thanks.

junrao commented on core/src/test/java/kafka/server/ClientMetricsManagerTest.java:
```diff
@@ -0,0 +1,921 @@
+    }
+
+    @Test
+    public void testGetTelemetryWithoutSubscription() throws UnknownHostException {
```
Does this test provide more coverage than `testGetTelemetry()`?

apoorvmittal10 replied:
This test provides similar coverage as like GetTelemetry but validates the scenario when no subscription is yet created and telemetry API is invoked then the request shall succeed considering default properties i.e. subscriptionMap is empty.

junrao commented on core/src/test/java/kafka/server/ClientMetricsManagerTest.java:
```diff
@@ -0,0 +1,921 @@
+                responses.add(response);
+                lock.countDown();
+            } catch (UnknownHostException e) {
+                e.printStackTrace();
```
We should use logging instead of print. Ditto in other places.

apoorvmittal10 replied:
Done.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -30,17 +68,366 @@ public class ClientMetricsManager implements Closeable {
+        int telemetryMaxBytes, RequestContext requestContext, int throttleMs) {
+
+        Uuid clientInstanceId = request.data().clientInstanceId();
+        if (clientInstanceId == null || clientInstanceId == Uuid.ZERO_UUID) {
```
It seems that `clientInstanceId` can't be any in `Uuid.RESERVED`?

apoorvmittal10 replied:
I have moved the check to Uuid.RESERVED, done.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -30,17 +68,366 @@ public class ClientMetricsManager implements Closeable {
+    private void validateGetRequest(GetTelemetrySubscriptionsRequest request,
+        ClientMetricsInstance clientInstance, long timestamp) {
+
+        if (!clientInstance.maybeUpdateGetRequestEpoch(timestamp) && (clientInstance.lastKnownError() != Errors.UNKNOWN_SUBSCRIPTION_ID
```
Why only testing these two errors? What about other errors like `InvalidRequest`?

apoorvmittal10 replied:
UNKNOWN_SUBSCRIPTION_ID and UNSUPPORTED_COMPRESSION_TYPE are the only 2 exception types defined in KIP-714 which shall allow immediate retry of GetTelemetrySubscriptions request after an error from PushTelemetryRequest. For all other exception types either the API calls should be disallowed or should follow the regular push interval cycle.

junrao commented on core/src/main/java/kafka/metrics/ClientMetricsInstance.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    public boolean terminating() {
+        return terminating;
```
After the instance is terminated, when do we remove the instance form the in-memory state?

apoorvmittal10 replied:
The reason why I didn't remove the instance from in-memory cache as the subsequent requests from terminated client should be rejected. Having said that, the terminated client should also not remain in-memory forever hence the terminated client should be removed from cache as per the eviction policy of cache i.e. `MAX(60*1000, PushIntervalMs * 3) milliseconds`. This shall be inherently handled by the cache improvement task where we have some time based eviction policy.

junrao commented on core/src/test/java/kafka/metrics/ClientMetricsInstanceTest.java:
```diff
@@ -0,0 +1,104 @@
+        assertTrue(clientInstance.maybeUpdateGetRequestEpoch(System.currentTimeMillis()));
+        // sleep for 3 ms to ensure that the next request is accepted.
+        Thread.sleep(3);
+        // Second request should be accepted as time since last request is greater than the retry interval.
```
retry interval => push interval

apoorvmittal10 replied:
Done.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. Made a pass of all files. A few more comments.

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 commented:
Thanks for reviewing @junrao, I have updated the PR.

junrao commented on core/src/test/java/kafka/metrics/ClientMetricsInstanceTest.java:
```diff
@@ -0,0 +1,92 @@
+    @Test
+    public void testMaybeUpdateRequestEpochValid() {
+        // First request should be accepted.
+        assertTrue(clientInstance.maybeUpdateGetRequestEpoch(System.currentTimeMillis()));
```
Could we use MockTime instead of System time?

apoorvmittal10 replied:
As the class/methods I am testing do not have `Time` reference rather just tries to updates the epoch supplied hence I didn't see any value in using MockTime. However I did change the code and tests for ClientMetricsManager where these methods are invoked from, I started using Time in ClientMetricsManager and corresponding MockTime in ClientMetricsManagerTest which eliminated the use of `Thread.sleep`.

Please let me know if I am missing anything here.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/PushTelemetryRequest.java:
```diff
@@ -60,17 +62,47 @@ public PushTelemetryRequest(PushTelemetryRequestData data, short version) {
+         In this case, the throttleTimeMs should not be included in the response to avoid the
+         muting of the client channel for all the other requests.
+        */
+        if (Errors.THROTTLING_QUOTA_EXCEEDED != errors) {
```
I noticed that sometimes we do set `throttleTimeMs` with an error. However, this only happens in KafkaApis with the generic request throttling logic. So, for now, I'd recommend that we get rid of this check here and rely on the caller to set `throttleTimeMs` properly. Specifically, for all `errorResponse(int throttleTimeMs, Errors errors)` and `getErrorResponse(int throttleTimeMs, Throwable e)` calls in `ClientMetricsManager`, we should pass in 0 for `throttleTimeMs`.

apoorvmittal10 replied:
I have removed the conditional llogic in this class and set `0` as the throttle time in case of error in ClientMetricsManager. Done.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -30,17 +69,376 @@ public class ClientMetricsManager implements Closeable {
+        List<String> clientMatchPattern = configs.getList(ClientMetricsConfigs.CLIENT_MATCH_PATTERN);
+
+        /*
+         Update last subscription updated time to current time to indicate that there is a change
```
The comment seems outdated since we don't update the time now.

apoorvmittal10 replied:
Done.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -30,17 +69,376 @@ public class ClientMetricsManager implements Closeable {
+         in the subscription. This will be used to determine if the next telemetry request needs
+         to re-evaluate the subscription id as per changes subscriptions.
+        */
+        int version = this.subscriptionUpdateVersion.incrementAndGet();
```
What's the convention of using `this` when accessing a field? It seems that class sometimes uses `this` and some other times doesn't.

apoorvmittal10 replied:
Corrected it, `this` should be used at places where there is a local variable with same name else directly instance variables should be accessed.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -30,17 +69,376 @@ public class ClientMetricsManager implements Closeable {
+                    ClientMetricsConfigs.ALL_SUBSCRIBED_METRICS_CONFIG);
+                subscribedMetrics.addAll(info.metrics());
+                pushIntervalMs = Math.min(pushIntervalMs, info.intervalMs());
+                currentSubscriptionVersion = Math.max(currentSubscriptionVersion, info.version());
```
Hmm, if not all subscriptions match a client, `currentSubscriptionVersion` will always be below the global `subscriptionUpdateVersion` and it forces a new client instance to be recreated on every request, right? Could we just use the global `subscriptionUpdateVersion`?

apoorvmittal10 replied:
My bad, I have used the global `subscriptionUpdateVersion`

junrao commented on core/src/main/java/kafka/metrics/ClientMetricsInstance.java:
```diff
@@ -0,0 +1,123 @@
+    private final Set<String> metrics;
+    private final int pushIntervalMs;
+
+    private long lastGetRequestEpoch;
```
`lastGetRequestEpoch` => `lastGetRequestTimestamp`? Ditto for `lastPushRequestEpoch`.

apoorvmittal10 replied:
Done.

apoorvmittal10 committed:
Moved client manager initialization, addressed comments

apoorvmittal10 commented:
Thanks @junrao for leaving the comments, I have tried to address them.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -16,31 +16,421 @@
+
+    /**
+     * Computes the SubscriptionId as a unique identifier for a client instance's subscription set,
+     * the id is generated by calculating a CRC32 of the configured metrics subscriptions including
```
CRC32 => CRC32C

apoorvmittal10 replied:
Done.

junrao commented on core/src/test/java/kafka/server/ClientMetricsManagerTest.java:
```diff
@@ -0,0 +1,951 @@
+
+        assertEquals(Errors.NONE, response.error());
+
+        time.setCurrentTimeMs(time.milliseconds() + ClientMetricsTestUtils.DEFAULT_PUSH_INTERVAL_MS);
```
This could just be `time.sleep(ClientMetricsTestUtils.DEFAULT_PUSH_INTERVAL_MS)`. Ditto in a few other places.

apoorvmittal10 replied:
Done.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -16,31 +16,421 @@
+                ClientMetricsReceiverPlugin.instance().exportMetrics(requestContext, request);
+            } catch (Exception exception) {
+                clientInstance.lastKnownError(Errors.INVALID_RECORD);
+                return request.errorResponse(throttleMs, Errors.INVALID_RECORD);
```
`throttleMs` should be 0.

apoorvmittal10 replied:
Missed this, done.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -16,31 +16,421 @@
+        }
+
+        clientInstance.lastKnownError(Errors.NONE);
+        return new PushTelemetryResponse(new PushTelemetryResponseData().setThrottleTimeMs(throttleMs));
```
As discussed, we shouldn't set `throttleMs` here.

apoorvmittal10 replied:
As per the details mentioned earlier, that we shouldn't set `throttleMs` in case of errors from telemetry APIs else channel can be muted. But shouldn't we set `throttleMs` at all from telemetry APIs i.e. in case of successful response but broker actually needs to throttle the channel? I might be missing something here.


https://github.com/apache/kafka/pull/14699#discussion_r1391827581
https://github.com/apache/kafka/pull/14699#discussion_r1397714746

junrao replied:
Currently, we have a generic request quota (for CPU usage). It mutes the channel for `throttleMs` and sets no error code in the response.

We also have additional quotas for specific requests. For example, https://cwiki.apache.org/confluence/display/KAFKA/KIP-599 adds controller mutation rate quota. It mutes the channel for `throttleMs` and set THROTTLING_QUOTA_EXCEEDED  error in the response.

For the new throttling introduced in KIP-714, I guess the intention is to do sth similar to controller mutation rate quota. If we want to do that, we need to do the following.
1. Just setting `throttleMs` in the response won't mute the channel. We need to introduce a new type of `QuotaManager` in `QuotaManagers` and wire the handling of `pushTelemetry` request to the new `QuotaManager` so that it could use the generic framework for muting/unmuting the channel, updating the metric, etc.
2. The existing throttling is rate based. It would be useful to map the new throttling in KIP-714 to that model so that we could get consistent throttling related metrics.
3. `PushTelemetry` will also be subject to the request quota. So, we need a mechanism to combine the throttling time from two different quota managers. We can follow how `KafkaApis.handleProduceRequest` does that (since it needs to combine request quota and the produce quota).

apoorvmittal10 replied:
Thanks @junrao for explaining this. I have removed throttleMs from PushTelemetry response. Also created below Jira to add changes in QuotaManager. Resolving the comment.

https://issues.apache.org/jira/browse/KAFKA-15863

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. Left a few more comments.

apoorvmittal10 committed:
Updated files as per review comments

apoorvmittal10 commented:
Thanks for the review @junrao. I have updated the PR.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -16,31 +16,420 @@
  */
 package kafka.server;
 
+import kafka.metrics.ClientMetricsConfigs;
```
With https://issues.apache.org/jira/browse/KAFKA-15853, we now have a new server module. So, we want to move this class and all classes in core/src/java/metrics there. This could be done in a follow up jira.

apoorvmittal10 replied:
As I received merge conflict in the current PR hence merged the upstream changes which required to move the classes to write package. I have done this in current PR itself.

junrao commented on core/src/main/scala/kafka/server/metadata/DynamicConfigPublisher.scala:
```diff
@@ -103,9 +103,16 @@ class DynamicConfigPublisher(
+              dynamicConfigHandlers.get(ConfigType.ClientMetrics).foreach(metricsConfigHandler =>
+                try {
+                  info(s"Updating client metrics ${resource.name()} with new configuration : " +
+                    toLoggableProps(resource, props).mkString(","))
```
`toLoggableProps` by default treats the property as sensitive. Should we add the Client_Metrics type support in `KafkaConfig.loggableValue`?

apoorvmittal10 replied:
Done.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. A few more comments.

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into kip-714-ak-manager

apoorvmittal10 committed:
Resolving merge issue

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 committed:
Removed throttleMs from ClientMetricsManager

apoorvmittal10 commented:
Thanks @junrao for explaining the details. I have updated the PR and removed throttleMs from ClientMetricsManager. I have added a Jira to add respective throttling changes in QuotaManager.

apoorvmittal10 committed:
Fixing import control

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. LGTM. Are the 34 test failures related?

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into kip-714-ak-manager

apoorvmittal10 commented:
> @apoorvmittal10 : Thanks for the updated PR. LGTM. Are the 34 test failures related?

Thanks @junrao, the tests are not related and I ll trigger re-build to see if tests fail persists.

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into kip-714-ak-manager

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into kip-714-ak-manager

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into kip-714-ak-manager

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/common/protocol/Errors.java:
```diff
@@ -386,7 +388,9 @@ public enum Errors {
     UNSUPPORTED_ENDPOINT_TYPE(115, "This endpoint type is not supported yet.", UnsupportedEndpointTypeException::new),
-    UNKNOWN_CONTROLLER_ID(116, "This controller ID is not known.", UnknownControllerIdException::new);
+    UNKNOWN_CONTROLLER_ID(116, "This controller ID is not known.", UnknownControllerIdException::new),
+    UNKNOWN_SUBSCRIPTION_ID(117, "Client sent a push telemetry request with an invalid or outdated subscription ID", UnknownSubscriptionIdException::new),
```
Tiny change request. Please put a full stop at the end of the error strings. `"Client sent a push telemetry request with an invalid or outdated subscription ID."`.

apoorvmittal10 replied:
Done, thanks for pointing out.

apoorvmittal10 committed:
Adding . in the error message

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into kip-714-ak-manager

apoorvmittal10 commented:
@junrao @AndrewJSchofield Thanks for the LGTM and approving the PR. Below is the test run status, none of the failing tests are related to the changes in the PR but I have tried to debug further. Among 20, 17 have already been reported as flaky test in jira, for remaining 3 I tried to locally reproduce but couldn't hence created `flaky test` for same. Can we merge the PR? 

New failing - 20

`Already existing jira for flaky test`: Build / JDK 11 and Scala 2.13 / testResetSinkConnectorOffsetsZombieSinkTasks – org.apache.kafka.connect.integration.OffsetsApiIntegrationTest
1m 2s: https://issues.apache.org/jira/browse/KAFKA-15524

`Already existing jira for flaky test`: Build / JDK 11 and Scala 2.13 / testRackAwareRangeAssignor(String).quorum=kraft – integration.kafka.server.FetchFromFollowerIntegrationTest
8s: https://issues.apache.org/jira/browse/KAFKA-15020

`Already existing jira for flaky test`: Build / JDK 11 and Scala 2.13 / testRackAwareRangeAssignor(String).quorum=kraft – integration.kafka.server.FetchFromFollowerIntegrationTest
6s: https://issues.apache.org/jira/browse/KAFKA-15020

`Already existing jira for flaky test`: Build / JDK 11 and Scala 2.13 / testTaskRequestWithOldStartMsGetsUpdated() – org.apache.kafka.trogdor.coordinator.CoordinatorTest
2m 0s: https://issues.apache.org/jira/browse/KAFKA-15760

`Already existing jira for flaky test`: Build / JDK 21 and Scala 2.13 / testReplicateSourceDefault() – org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest
2m 0s: https://issues.apache.org/jira/browse/KAFKA-15292

`Already existing jira for flaky test`: Build / JDK 21 and Scala 2.13 / testConsumptionWithBrokerFailures() – kafka.api.ConsumerBounceTest
31s: https://issues.apache.org/jira/browse/KAFKA-15146

`Already existing jira for flaky test`: Build / JDK 21 and Scala 2.13 / testDescribeClusterRequestIncludingClusterAuthorizedOperations(String).quorum=kraft – kafka.server.DescribeClusterRequestTest
4s: https://issues.apache.org/jira/browse/KAFKA-15759

`Already existing jira for flaky test`: Build / JDK 21 and Scala 2.13 / testSendOffsetsWithGroupId(String).quorum=zk – org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest
1m 15s: https://issues.apache.org/jira/browse/KAFKA-15772

`Created Jira, cannot reproduce locally (also exists other flaky tests from same class, have already ben reported by others)`: Build / JDK 21 and Scala 2.13 / shouldQuerySpecificStalePartitionStores() – org.apache.kafka.streams.integration.StoreQueryIntegrationTest
4s: Another test in same class flaky: https://issues.apache.org/jira/browse/KAFKA-9897, created jira: https://issues.apache.org/jira/browse/KAFKA-15896

`Already existing jira for flaky test`: Build / JDK 17 and Scala 2.13 / testReplicateSourceDefault() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationBaseTest
2m 23s: https://issues.apache.org/jira/browse/KAFKA-15699

`Already existing jira for flaky test`: Build / JDK 17 and Scala 2.13 / testMultiWorkerRestartOnlyConnector – org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest
2m 33s: https://issues.apache.org/jira/browse/KAFKA-15675

`Already existing jira for flaky test`: Build / JDK 17 and Scala 2.13 / testDescribeTokenForOtherUserPasses(String).quorum=kraft – kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest
9s: https://issues.apache.org/jira/browse/KAFKA-15411

`Created flaky test jira, cannot reproduce locally`: Build / JDK 17 and Scala 2.13 / testWrongIncarnationId() – kafka.server.ControllerRegistrationManagerTest
30s: https://issues.apache.org/jira/browse/KAFKA-15897

`Created flaky test jira, cannot reproduce locally`: Build / JDK 17 and Scala 2.13 / testFenceMultipleBrokers() – org.apache.kafka.controller.QuorumControllerTest
43s: https://issues.apache.org/jira/browse/KAFKA-15898

`Already existing jira for flaky test`: Build / JDK 8 and Scala 2.12 / testSyncTopicConfigs() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationBaseTest
2m 33s: https://issues.apache.org/jira/browse/KAFKA-14971

`Already existing jira for flaky test`: Build / JDK 8 and Scala 2.12 / [2] quorum=kraft, isIdempotenceEnabled=false – kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest
10s: https://issues.apache.org/jira/browse/KAFKA-15411

`Already existing jira for flaky test`: Build / JDK 8 and Scala 2.12 / testDescribeClusterRequestExcludingClusterAuthorizedOperations(String).quorum=kraft – kafka.server.DescribeClusterRequestTest
4s: https://issues.apache.org/jira/browse/KAFKA-15419

`Already existing jira for flaky test`: Build / JDK 8 and Scala 2.12 / shouldWriteLatestOffsetsToCheckpointOnShutdown[at_least_once, processing threads = true] – org.apache.kafka.streams.integration.EosIntegrationTest
1m 25s: https://issues.apache.org/jira/browse/KAFKA-15690

`Already existing jira for flaky test`: Build / JDK 8 and Scala 2.12 / shouldWriteLatestOffsetsToCheckpointOnShutdown[exactly_once, processing threads = true] – org.apache.kafka.streams.integration.EosIntegrationTest
1m 21s: https://issues.apache.org/jira/browse/KAFKA-15690

`Already existing jira for flaky test`: Build / JDK 8 and Scala 2.12 / shouldWriteLatestOffsetsToCheckpointOnShutdown[exactly_once_v2, processing threads = true] – org.apache.kafka.streams.integration.EosIntegrationTest: https://issues.apache.org/jira/browse/KAFKA-15690

junrao commented:
Just merged https://github.com/apache/kafka/pull/14632. Triggering another test run to make sure there are no new issues.

apoorvmittal10 commented:
> Just merged #14632. Triggering another test run to make sure there are no new issues.

Thanks @junrao.

junrao commented:
@apoorvmittal10 : It seems that the build for JDK 17 and Scala 2.13 didn't complete.

apoorvmittal10 commented:
> @apoorvmittal10 : It seems that the build for JDK 17 and Scala 2.13 didn't complete.

Thanks @junrao. Strange to see the failure in. `:clients:test`. I can see below unrelated failure in logs, retriggered build.

```
> Task :clients:test
org.apache.kafka.common.network.SelectorTest.testConnectionsByClientMetric() failed, log available in /home/jenkins/jenkins-agent/712657a4/workspace/Kafka_kafka-pr_PR-14699/clients/build/reports/testOutput/org.apache.kafka.common.network.SelectorTest.testConnectionsByClientMetric().test.stdout

Gradle Test Run :clients:test > Gradle Test Executor 13 > SelectorTest > testConnectionsByClientMetric() FAILED
    java.util.concurrent.TimeoutException: testConnectionsByClientMetric() timed out after 240 seconds
        at org.junit.jupiter.engine.extension.TimeoutExceptionFactory.create(TimeoutExceptionFactory.java:29)
        at org.junit.jupiter.engine.extension.SameThreadTimeoutInvocation.proceed(SameThreadTimeoutInvocation.java:58)
        at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
        at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
        at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:218)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:214)

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into kip-714-ak-manager

apoorvmittal10 commented:
Hi @junrao, I have triaged the test cases as per the latest run.


`Existing`: Build / JDK 17 and Scala 2.13 / testMultiWorkerRestartOnlyConnector – org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest
2m 21s: https://issues.apache.org/jira/browse/KAFKA-15675

`Existing`: Build / JDK 17 and Scala 2.13 / testRackAwareRangeAssignor(String).quorum=zk – integration.kafka.server.FetchFromFollowerIntegrationTest
39s: https://issues.apache.org/jira/browse/KAFKA-15020

`Existing`: Build / JDK 17 and Scala 2.13 / testCoordinatorFailover(String, String).quorum=zk.groupProtocol=generic – kafka.api.SslConsumerTest
8s: https://issues.apache.org/jira/browse/KAFKA-15920

`Existing`: Build / JDK 17 and Scala 2.13 / testDescribeClusterRequestExcludingClusterAuthorizedOperations(String).quorum=kraft – kafka.server.DescribeClusterRequestTest
3s: https://issues.apache.org/jira/browse/KAFKA-15419

`Existing`: Build / JDK 17 and Scala 2.13 / testFenceMultipleBrokers() – org.apache.kafka.controller.QuorumControllerTest
51s: https://issues.apache.org/jira/browse/KAFKA-15898

`Existing`: Build / JDK 21 and Scala 2.13 / testSyncTopicConfigs() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationBaseTest
1m 42s: https://issues.apache.org/jira/browse/KAFKA-15523

`Existing`: Build / JDK 21 and Scala 2.13 / testTaskRequestWithOldStartMsGetsUpdated() – org.apache.kafka.trogdor.coordinator.CoordinatorTest
2m 0s: https://issues.apache.org/jira/browse/KAFKA-15760

`Existing`: Build / JDK 11 and Scala 2.13 / testSyncTopicConfigs() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationTransactionsTest
3m 44s: https://issues.apache.org/jira/browse/KAFKA-15523

`Triaged`: Build / JDK 11 and Scala 2.13 / testRestartReplication() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationTransactionsTest
3m 21s: https://issues.apache.org/jira/browse/KAFKA-15933

`Triaged`: Build / JDK 11 and Scala 2.13 / testMultiConsumerStickyAssignor(String, String).quorum=kraft+kip848.groupProtocol=generic – kafka.api.PlaintextConsumerTest
2m 12s: https://issues.apache.org/jira/browse/KAFKA-15934

`Existing`: Build / JDK 8 and Scala 2.12 / testSyncTopicConfigs() – org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest
4m 25s: https://issues.apache.org/jira/browse/KAFKA-14971

`Triaged`: Build / JDK 8 and Scala 2.12 / testRestartReplication() – org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest
2m 8s: https://issues.apache.org/jira/browse/KAFKA-15935

`Existing`: Build / JDK 8 and Scala 2.12 / testNoCheckpointsIfNoRecordsAreMirrored() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationBaseTest
2m 59s: https://issues.apache.org/jira/browse/KAFKA-15699

`Existing`: Build / JDK 8 and Scala 2.12 / testSyncTopicConfigs() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationTransactionsTest
3m 37s: https://issues.apache.org/jira/browse/KAFKA-15523

`Existing`: Build / JDK 8 and Scala 2.12 / testMultiWorkerRestartOnlyConnector – org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest
2m 28s: https://issues.apache.org/jira/browse/KAFKA-15761

`Existing`: Build / JDK 8 and Scala 2.12 / testMultiWorkerRestartOnlyConnector – org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest
2m 39s: https://issues.apache.org/jira/browse/KAFKA-15761

`Triaged`: Build / JDK 8 and Scala 2.12 / testGetActiveTopics – org.apache.kafka.connect.integration.ConnectorTopicsIntegrationTest
1m 50s: https://issues.apache.org/jira/browse/KAFKA-15936

`Triaged`: Build / JDK 8 and Scala 2.12 / testTopicTrackingResetIsDisabled – org.apache.kafka.connect.integration.ConnectorTopicsIntegrationTest
2m 37s: https://issues.apache.org/jira/browse/KAFKA-15936

`Triaged`: Build / JDK 8 and Scala 2.12 / testTopicTrackingIsDisabled – org.apache.kafka.connect.integration.ConnectorTopicsIntegrationTest
3m 4s: https://issues.apache.org/jira/browse/KAFKA-15936

`Existing`: Build / JDK 8 and Scala 2.12 / testConnectorReconfiguration – org.apache.kafka.connect.integration.ExactlyOnceSourceIntegrationTest
1m 42s: https://issues.apache.org/jira/browse/KAFKA-14901

`Existing`: Build / JDK 8 and Scala 2.12 / testSeparateOffsetsTopic – org.apache.kafka.connect.integration.ExactlyOnceSourceIntegrationTest
3m 35s: https://issues.apache.org/jira/browse/KAFKA-14089

`Triaged`: Build / JDK 8 and Scala 2.12 / testTopicDeletion(String).quorum=kraft – kafka.admin.RemoteTopicCrudTest
1m 30s: https://issues.apache.org/jira/browse/KAFKA-15937

`Triaged`: Build / JDK 8 and Scala 2.12 / testCreateRemoteTopicWithValidRetentionTime(String).quorum=kraft – kafka.admin.RemoteTopicCrudTest
2m 17s: https://issues.apache.org/jira/browse/KAFKA-15938

`Triaged`: Build / JDK 8 and Scala 2.12 / testInvalidAlterConfigs(String).quorum=kraft – kafka.api.AdminClientWithPoliciesIntegrationTest
2m 2s: https://issues.apache.org/jira/browse/KAFKA-15939

`Existing`: Build / JDK 8 and Scala 2.12 / testDescribeClusterRequestExcludingClusterAuthorizedOperations(String).quorum=kraft – kafka.server.DescribeClusterRequestTest
3s: https://issues.apache.org/jira/browse/KAFKA-15419

`Triaged`: Build / JDK 8 and Scala 2.12 / testLogCleanerConfig(String).quorum=kraft – kafka.server.DynamicBrokerReconfigurationTest
41s: https://issues.apache.org/jira/browse/KAFKA-15940

`Existing`: Build / JDK 8 and Scala 2.12 / shouldHaveSamePositionBoundActiveAndStandBy – org.apache.kafka.streams.integration.ConsistencyVectorIntegrationTest
22s: https://issues.apache.org/jira/browse/KAFKA-15770

`Existing`: Build / JDK 8 and Scala 2.12 / shouldWriteLatestOffsetsToCheckpointOnShutdown[exactly_once_v2, processing threads = true] – org.apache.kafka.streams.integration.EosIntegrationTest
2m 11s: https://issues.apache.org/jira/browse/KAFKA-15690

`Existing`: Build / JDK 8 and Scala 2.12 / shouldUpgradeFromEosAlphaToEosV2[false] – org.apache.kafka.streams.integration.EosV2UpgradeIntegrationTest
1m 47s: https://issues.apache.org/jira/browse/KAFKA-15095

`Existing`: Build / JDK 8 and Scala 2.12 / shouldUpgradeFromEosAlphaToEosV2[true] – org.apache.kafka.streams.integration.EosV2UpgradeIntegrationTest
4m 31s: https://issues.apache.org/jira/browse/KAFKA-15095

`Triaged`: Build / JDK 8 and Scala 2.12 / shouldRestoreNullRecord() – org.apache.kafka.streams.integration.RestoreIntegrationTest
1m 5s: https://issues.apache.org/jira/browse/KAFKA-15941

`Existing`: Build / JDK 8 and Scala 2.12 / testTopicDeletion(String).quorum=kraft – org.apache.kafka.tools.TopicCommandIntegrationTest
4m 11s: https://issues.apache.org/jira/browse/KAFKA-15140

`Existing`: Build / JDK 8 and Scala 2.12 / testDescribeReportOverriddenConfigs(String).quorum=kraft – org.apache.kafka.tools.TopicCommandIntegrationTest: https://issues.apache.org/jira/browse/KAFKA-15140

apoorvmittal10 commented:
@junrao The tests failures are not related to the changes.

junrao reviewed (approved):
Thanks for the analysis, @apoorvmittal10. The PR LGTM.

chia7712 commented on server/src/main/java/org/apache/kafka/server/ClientMetricsManager.java:
```diff
@@ -16,31 +16,417 @@
+            return request.getErrorResponse(0, exception);
+        } finally {
+            // Update the client instance with the latest push request parameters.
+            clientInstance.terminating(request.data().terminating());
```
We noticed an edge case and have opened #23094 to discuss it. It would be great to get your review :)
